File: lib/java/src/main/java/io/github/brenbar/uapi/internal/binary/PackList.java
package io.github.brenbar.uapi.internal.binary;

import static io.github.brenbar.uapi.internal.binary.Pack.pack;
import static io.github.brenbar.uapi.internal.binary.PackMap.packMap;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.msgpack.jackson.dataformat.MessagePackExtensionType;

public class PackList {

    public static final byte PACKED_BYTE = (byte) 17;

    static List<Object> packList(List<Object> list) {
        if (list.isEmpty()) {
            return list;
        }

        final var packedList = new ArrayList<Object>();
        final var header = new ArrayList<Object>();

        packedList.add(new MessagePackExtensionType(PACKED_BYTE, new byte[0]));

        header.add(null);

        packedList.add(header);

        final var keyIndexMap = new HashMap<Integer, BinaryPackNode>();
        try {
            for (final var e : list) {
                if (e instanceof final Map<?, ?> m) {
                    final var row = packMap(m, header, keyIndexMap);

                    packedList.add(row);
                } else {
                    // This list cannot be packed, abort
                    throw new CannotPack();
                }
            }
            return packedList;
        } catch (final CannotPack ex) {
            final var newList = new ArrayList<Object>();
            for (final var e : list) {
                newList.add(pack(e));
            }
            return newList;
        }
    }
}

File: lib/py/telepact/internal/binary/PackList.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import TYPE_CHECKING
from threading import Lock

from msgpack import ExtType
from .CannotPack import CannotPack
from .PackMap import pack_map

if TYPE_CHECKING:
    from ...internal.binary.BinaryPackNode import BinaryPackNode


PACKED_BYTE = 17
PACKED_EXT = ExtType(PACKED_BYTE, b'')
_PACK = None
_PACK_LOCK = Lock()


def _get_pack():
    global _PACK
    if _PACK is None:
        with _PACK_LOCK:
            if _PACK is None:
                from .Pack import pack as _pack
                _PACK = _pack
    return _PACK


def pack_list(lst: list[object]) -> list[object]:
    pack = _get_pack()

    if not lst:
        return lst

    packed_list: list[object] = []
    header: list[object] = []

    packed_list.append(PACKED_EXT)

    header.append(None)

    packed_list.append(header)

    key_index_map: dict[int, BinaryPackNode] = {}
    try:
        for item in lst:
            if type(item) is dict:
                row = pack_map(item, header, key_index_map)
                packed_list.append(row)
            else:
                # This list cannot be packed, abort
                raise CannotPack()
        return packed_list
    except CannotPack:
        new_list = [pack(item) for item in lst]
        return new_list
